components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp

```diff
@@ -27,329 +41,271 @@ void Nudging::set_grids(const std::shared_ptr<const GridsManager> grids_manager)
+  m_time_interp.perform_time_interpolation(ts);
+
+  // Process data and nudge the atmosphere state
+  const auto& p_mid_v     = get_field_in("p_mid").get_view<const mPack**>();
+  view_Nd<mPack,2> p_mid_ext_p;
+  view_Nd<mPack,1> p_mid_ext_1d;
+  if (m_src_pres_type == DYNAMIC) {
+    const auto& p_mid_ext   = get_helper_field("p_mid_ext").get_view<mPack**>();
+    p_mid_ext_p = view_Nd<mPack,2>(reinterpret_cast<mPack*>(p_mid_ext.data()),
+  				     m_num_cols,m_num_src_levs);
+  } else if (m_src_pres_type == STATIC) {
+    p_mid_ext_1d   = get_helper_field("p_mid_ext").get_view<mPack*>();
+  }
+  for (auto name : m_fields_nudge) {
+    auto atm_state_field = get_field_out(name);
+    auto int_state_field = get_helper_field(name);
+    auto ext_state_field = get_helper_field(name+"_ext").get_view<mPack**>();
+    auto atm_state_view  = atm_state_field.get_view<mPack**>();  // TODO: Right now assume whatever field is defined on COLxLEV
+    auto int_state_view  = int_state_field.get_view<mPack**>();
+    view_Nd<mMask,2> int_mask_view("",m_num_cols,m_num_levs); // Track mask of interpolated values
+    const view_Nd<mPack,2> ext_state_view(reinterpret_cast<mPack*>(ext_state_field.data()),
+                                          m_num_cols,m_num_src_levs);
+    // Masked values in the source data can lead to strange behavior in the vertical interpolation.
+    // We pre-process the data and map any masked values (sometimes called "filled" values) to the
+    // nearest un-masked value.
+    Real var_fill_value;
+    // WARNING!!! Assumes that all datafiles use the same FillValue.  It is unlikely that a user will use a mismatch of files
+    // with different defined FillValues, but if they do, this loop won't catch the change.
+    scorpio::get_variable_metadata(m_datafiles[0],name,"_FillValue",var_fill_value);
+    const int num_cols           = int_state_view.extent(0);
+    const int num_vert_packs     = int_state_view.extent(1);
+    const auto policy = ESU::get_default_team_policy(num_cols, num_vert_packs);
+    Kokkos::parallel_for("correct_for_masked_values", policy,
+       	       KOKKOS_LAMBDA(MemberType const& team) {
       const int icol = team.league_rank();
-      auto T_mid_1d = ekat::subview(T_mid,icol);
-      auto T_mid_ext_1d = ekat::subview(T_mid_ext,icol);
-      auto u_1d = ekat::subview(u,icol);
-      auto u_ext_1d = ekat::subview(u_ext,icol);
-      auto v_1d = ekat::subview(v,icol);
-      auto v_ext_1d = ekat::subview(v_ext,icol);
+      auto int_mask_view_1d  = ekat::subview(int_mask_view,icol);
+      auto int_state_view_1d = ekat::subview(int_state_view,icol);
+      Real fill_value;
+      int  fill_idx = -1;
+      // Scan top to surf and backfill all values near TOM that are masked.
+      for (int kk=0; kk<m_num_levs; ++kk) {
+        const auto ipack = kk / mPack::n;
+	const auto iidx  = kk % mPack::n;
+        // Check if this index is masked
+	if (!(int_state_view_1d(ipack)[iidx]==var_fill_value)) {
```

Comment:
Two things:
- instead of `if (!(a==b))`, you can just do `if (a!=b)`, which is more readable.
- I don't get the logic: if the entry is not the fill value, then you set fill_value=entry...doesn't that set the view to a _valid_ value in the entries that were previously filled?